src/librustc_trans/trans/debuginfo/metadata.rs

```diff
@@ -796,12 +796,31 @@ pub fn type_metadata<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
             }
         }
         ty::TyBareFn(_, ref barefnty) => {
-            subroutine_type_metadata(cx, unique_type_id, &barefnty.sig, usage_site_span)
+            let fn_metadata = subroutine_type_metadata(cx,
```

Comment:
The `match` checks whether `subroutine_type_metadata` happened to already create the metadata that we're trying to create, i.e. whether it  recursively called `type_metadata` for this type again.

Thinking about it, I feel like this isn't actually possible for function types, so the `match` could be dropped. Is that correct?